src/utils/symlinkPackageDependencies.js

```diff
@@ -0,0 +1,219 @@
+
+    if (!actualBinFileRelative) {
+      throw new BoltError(`${binName} is not a symlink`);
+    }
+
+    const actualBinFile = path.join(
+      project.pkg.nodeModulesBin,
+      actualBinFileRelative
+    );
+
+    if (!pathIsInside(actualBinFile, project.pkg.nodeModules)) {
+      throw new BoltError(
+        `${binName} is linked to a location outside of project node_modules: ${actualBinFileRelative}`
+      );
+    }
+
+    // To find the name of the dep that created the bin we'll get its path from node_modules and
+    // use the first one or two parts (two if the package is scoped)
+    const binFileRelativeToNodeModules = path.relative(
+      project.pkg.nodeModules,
+      actualBinFile
+    );
+    const pathParts = binFileRelativeToNodeModules.split(path.sep);
+    let pkgName = pathParts[0];
+
+    if (pkgName.startsWith('@')) {
+      pkgName += '/' + pathParts[1];
+    }
+
+    let workspaceBinPath = path.join(pkg.nodeModulesBin, binName);
+
+    symlinksToCreate.push({
+      src: binPath,
+      dest: workspaceBinPath,
+      type: 'exec'
+    });
+  }
+
+  /*****************************************************************
+   * Calculate all the internal bin files that need to be symlinked *
+  ******************************************************************/
+
+  // TODO: Same as above, we should really be making sure we get all the transitive bins as well
+
+  for (let dependency of internalDeps) {
+    const depWorkspace = dependencyGraph.get(dependency) || {};
+    const depBinFiles =
+      depWorkspace.pkg &&
+      depWorkspace.pkg.config &&
+      depWorkspace.pkg.config.getBin();
+
+    if (!depBinFiles) {
+      continue;
+    }
+
+    if (!includes(dependencies, dependency)) {
+      // dependency is not one we are supposed to symlink right now
+      continue;
+    }
+
+    if (typeof depBinFiles === 'string') {
+      // package may be scoped, name will only be the second part
+      const binName = dependency.split('/').pop();
+      const src = path.join(depWorkspace.pkg.dir, depBinFiles);
+      const dest = path.join(pkg.nodeModulesBin, binName);
+
+      symlinksToCreate.push({ src, dest, type: 'exec' });
+      continue;
+    }
+
+    for (let [binName, binPath] of Object.entries(depBinFiles)) {
+      const src = path.join(depWorkspace.pkg.dir, String(binPath));
+      const dest = path.join(pkg.nodeModulesBin, binName);
+
+      symlinksToCreate.push({ src, dest, type: 'exec' });
+    }
+  }
+
+  /**********************************
+   * Create directories and symlinks *
+  ***********************************/
+
+  await yarn.run(pkg, 'preinstall');
+
+  await Promise.all(
+    directoriesToCreate.map(dirName => {
+      return fs.mkdirp(dirName);
+    })
+  );
+
+  await Promise.all(
+    symlinksToCreate.map(async ({ src, dest, type }) => {
+      await fs.symlink(src, dest, type);
+    })
+  );
+
+  await yarn.run(pkg, 'postinstall');
+  await yarn.run(pkg, 'prepublish');
```

Comment:
Should also add `prepare`

- https://github.com/npm/npm/blob/latest/doc/misc/npm-scripts.md
- https://github.com/yarnpkg/yarn/pull/3004